finality-aleph/src/data_io.rs

```diff
@@ -359,82 +1011,228 @@ where
-    let mut prev_best_number: NumberFor<B> = proposed_block.lock().number;
-    loop {
-        let delay = futures_timer::Delay::new(Duration::from_millis(REFRESH_INTERVAL));
-        tokio::select! {
-            _ = delay => {
-                let new_best_header = select_chain
-                    .best_chain()
-                    .await
-                    .expect("No best chain");
-                if *new_best_header.number() <= max_block_num {
-                    *proposed_block.lock() = AlephData::new(new_best_header.hash(), *new_best_header.number());
-                } else {
-                    // we check if prev_best_header is an ancestor of new_best_header:
-                    if proposed_block.lock().number < max_block_num {
-                        let reduced_header = reduce_header_to_num(client.clone(), new_best_header.clone(), max_block_num);
-                        *proposed_block.lock() = AlephData::new(reduced_header.hash(), *reduced_header.number());
-                    } else {
-                        let reduced_header = reduce_header_to_num(client.clone(), new_best_header.clone(), prev_best_number);
-                        if reduced_header.hash() != prev_best_hash {
-                            // the new best block is not a descendant of previous best
-                            let reduced_header = reduce_header_to_num(client.clone(), new_best_header.clone(), max_block_num);
-                            *proposed_block.lock() = AlephData::new(reduced_header.hash(), *reduced_header.number());
-                        }
-                    }
-                }
-                prev_best_hash = new_best_header.hash();
-                prev_best_number = *new_best_header.number();
+    pub(crate) fn new(
+        select_chain: SC,
+        client: Arc<C>,
+        session_boundaries: SessionBoundaries<B>,
+        metrics: Option<Metrics<<B::Header as HeaderT>::Hash>>,
+    ) -> (Self, DataProvider<B>) {
+        let initial_data = AlephData::Empty;
+        let data_to_propose = Arc::new(Mutex::new(initial_data));
+        (
+            ChainTracker {
+                select_chain,
+                client,
+                data_to_propose: data_to_propose.clone(),
+                session_boundaries,
+                best_block_in_session: None,
+                prev_chain_info: None,
+            },
+            DataProvider {
+                data_to_propose,
+                metrics,
+            },
+        )
+    }
+
+    fn update_data(&mut self) {
+        // We read the best_block_in_session field and the highest_finalized block from the client and compute
+        // the corresponding `AlephData<B>` in `data_to_propose` for AlephBFT. To not recompute this many
+        // times we remember these "inputs" in `prev_chain_info` and upon match we leave the old value
+        // of `data_to_propose` unaffected.
+        let best_block = match &self.best_block_in_session {
```

Comment:
It looks rather mysterious that this seemingly unchanging field should have such an effect. I would suggest moving the logic that updates this field (currently in the `loop` in `run`) to a method, preferably one that returns the thing that we want to get here, to make this more readable. It might also be a good idea to more explicitly split this whole object into two, to make it clearer that these two bits of logic (updating `best_block` and `update_data`) are separate (are they?).

In general I have a problem with the code here being a tad messy, that's why I lashed out at the optimizations possibly unfairly. :/